---
title: "Methylation efficiency of enzymes"
author: "Kasit Chatsirisupachai and Guido Barzaghi"
date: '2025-07-01'
output:
  html_document: default
  pdf_document: default
---

```{r setup, include=FALSE}
library(here)
knitr::opts_knit$set(root.dir = here())
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = TRUE)
```

```{r check-wd}
getwd()
```


# Introduction
Single-molecule footprinting (SMF) is a technique to detect protein-DNA bindings at a single-molecule resolution. This technique leverages the ability of DNA methyltransferases, M.CviPI and M.SssI, to methylate cytosines at the accessible regions in the GpC and CpG context, respectively. In order to ensure an unbiased quantification of protein-DNA contacts, methylation efficiency should be uniform across all possible sequence contexts. In this workshop, we will examine M.CviPI and M.SssI methylation efficiencies in naked lambda DNA as a function of enzyme concentration and nucleotide contexts.


## Data
To footprint the DNA, Naked lambda DNA was treated with either M.CviPI (GpC) and M.SssI (CpG) in various concentrations (0, 0.2, 0.7, 0.75, 1.2, 1.5, 2, 3, 4, and 10 units/ug).
The sequencing was performed on Illumina MiSeq.

## Libraries
```{r}
library(QuasR)
library(SingleMoleculeFootprinting)
library(BSgenome)
library(tidyverse)
library(readr)
```

## Alignment!
We will first create QuasR input file, which will be used by QuasR package to locate the fastq files and sample name.
As the data was sequenced in paired-end, our QuasR input file will contain three columns: 1) file name for the read 1 fastq, 2) file name for the read 2 fastq, and 3) sample names.
For this example, the Qinput file was already generated `/analysis/example_1/Qinput_lambda_for_alignment.txt`, but need to be modified to add the right path to the file names. All necessary files for this example are in the folder `/data/dataset_1_lambdaDNA`

```{r}
Qinput <- "./analysis/example_1/Qinput_lambda_for_alignment.txt"
Qinput_df <- read.delim(Qinput)
print(Qinput_df)

# vector of sample names
sample_names <- Qinput_df$SampleName

# vector of fastq file names
read1_files <- Qinput_df$FileName1
read2_files <- Qinput_df$FileName2

# full paths
read1_full_paths <- here("data", "dataset_1_lambdaDNA", read1_files)
read2_full_paths <- here("data", "dataset_1_lambdaDNA", read2_files)

Qinput_df_new <- data.frame(FileName1 = read1_full_paths, 
                            FileName2 = read2_full_paths, 
                            SampleName = sample_names)

# write new Qinput file
write.table(Qinput_df_new, file = here("analysis", "example_1", "Qinput_lambda_for_alignment_full_path.txt"), sep = "\t", row.names = FALSE, quote = FALSE)
```

Now it's time to perform alignment. We will use QuasR package for this, which under the hood will run rbowtie.
```{r}
# load new Qinput
Qinput <- "./analysis/example_1/Qinput_lambda_for_alignment_full_path.txt"

QuasRprj <- QuasR::qAlign(sampleFile = Qinput, 
                          genome = "./data/dataset_1_lambdaDNA/LambdaGenome2.fasta", 
                          aligner = "Rbowtie",
                          paired = "fr",
                          bisulfite = "undir",
                          alignmentParameter="-e 70 -X 1000 -k 2 --best --strata",
                          alignmentsDir = "./data/dataset_1_lambdaDNA/test_alignment",
                          projectName = "test_alignment")
QuasRprj@aligner <- "Rbowtie"
```

Once it is done, you should have two BAM files. But there are a lot more samples, so for the sake of time, I already provided BAM files for all the conditions we will be working with.


## Working with provided BAM files
We will, again, first create QuasR input file, which will be used by QuasR package to locate the BAM files and sample name.
QuasR input file contains two columns: 1) BAM file names, and 2) sample names.
All BAM files for this example are in the folder `/data/dataset_1_lambdaDNA`
For this example, the Qinput file was already generated `/analysis/example_1/Qinput_lambda.txt`, but need to be modified to add the right path to the file names.
```{r}
Qinput <- "./analysis/example_1/Qinput_lambda.txt"
Qinput_df <- read.delim(Qinput)
print(head((Qinput_df)))

# vector of sample names
sample_names <- Qinput_df$SampleName

# vector of BAM file names
bam_files <- Qinput_df$FileName

# full paths
full_paths <- here("data", "dataset_1_lambdaDNA", bam_files)

Qinput_df_new <- data.frame(FileName = full_paths, SampleName = sample_names)

# write new Qinput file
write.table(Qinput_df_new, file = here("analysis", "example_1", "Qinput_lambda_full_path.txt"), sep = "\t", row.names = FALSE, quote = FALSE)
```

Load new Qinput
```{r}
Qinput <- "./analysis/example_1/Qinput_lambda_full_path.txt"
```

## Load reference genome
```{r}
LambdaGenomeSeq = readDNAStringSet(filepath = "./data/dataset_1_lambdaDNA/LambdaGenome2.fasta")
names(LambdaGenomeSeq) = "J02459.1"
LambdaGenomeSeq
```

## Alignment to the reference genome using QuasR
The files were already aligned. So, this step will only for checking the BAM files and create a QuasR project object.
```{r}
QuasRprj <- QuasR::qAlign(sampleFile = Qinput, 
                          genome = "./data/dataset_1_lambdaDNA/LambdaGenome2.fasta", 
                          paired = "fr",
                          bisulfite = "undir")
QuasRprj@aligner <- "Rbowtie"

QuasRprj
```

## Methylation Calling using QuasR
```{r}
Methylation = QuasR::qMeth(proj = QuasRprj, mode = "allC")

head(Methylation)
length(Methylation)
```

## Processing methylation calling output

### Separate GC and CG contexts
```{r}
GCs = FilterContextCytosines(MethGR = Methylation, genome = LambdaGenomeSeq, context = "GC")
print(head(GCs, 3))

CGs = FilterContextCytosines(MethGR = Methylation, genome = LambdaGenomeSeq, context = "CG")
print(head(CGs, 3))
```

### Collapse reads from + and - strands
```{r}
GCs_collapsed = CollapseStrands(MethGR = GCs, context = "GC")
head(GCs_collapsed, 3)

CGs_collapsed = CollapseStrands(MethGR = CGs, context = "HCG")
head(CGs_collapsed, 3)
```

### Filter cytosines by coverage (> 20 reads) and calculate methylation rate (methylated C / total C)
```{r}
GCs_filtered = CoverageFilter(MethGR = GCs_collapsed, thr = 20)
head(GCs_filtered, 3)

CGs_filtered = CoverageFilter(MethGR = CGs_collapsed, thr = 20)
head(CGs_filtered, 3)
```


### Extend sequence length around (methylated) cytosines. 
For the GC context, we add 2 nucleotides before C and 1 after C to get a NGCN pattern (C is at the third position). For CG context, we add 1 nucleotide before C and 2 after C to get a NCGN pattern (C is at the second position).
```{r}
start(GCs_filtered) = start(GCs_filtered) - 2
end(GCs_filtered) = end(GCs_filtered) + 1
start(CGs_filtered) = start(CGs_filtered) - 1
end(CGs_filtered) = end(CGs_filtered) + 2

AllCs = c(GCs_filtered, CGs_filtered)
head(AllCs)
```

Get sequence contexts of every cytosine tested
```{r}
KMERS = getSeq(LambdaGenomeSeq, AllCs)
head(KMERS, 3)
tail(KMERS, 3)

AllCs$Kmer_Context = KMERS
head(AllCs)
```


## Plotting the result
Let's plot median methylation efficiency by concentration of enzymes and Kmer contexts.

### Data frame manipulations
```{r}
AllCs %>%
  as_tibble() %>%
  filter(GenomicContext == "CG") %>%
  select(-seqnames, -end, -strand, -width) %>%
  gather(Measure, Value, -Kmer_Context, -GenomicContext, -start) %>%
  separate(Measure, into = c("Sample", "Measure"), sep = "_", remove = TRUE) %>%
  spread(Measure, Value) %>%
  filter(str_detect(Sample, "CG")) -> EDA_DF_CGs
head(EDA_DF_CGs)

AllCs %>%
  as_tibble() %>%
  filter(GenomicContext == "GC") %>%
  select(-seqnames, -end, -strand, -width) %>%
  gather(Measure, Value, -Kmer_Context, -GenomicContext, -start) %>%
  separate(Measure, into = c("Sample", "Measure"), sep = "_", remove = TRUE) %>%
  spread(Measure, Value)  %>%
  filter(str_detect(Sample, "GC"))-> EDA_DF_GCs
head(EDA_DF_GCs)
```


How to represent this result?
```{r}

```


