---
title: "Analysis of in-house generated GpC methylation (PEPC) vs commercial
  (NEB) enzymes"
author: "Kasit Chatsirisupachai"
date: "2024-06-02"
output:
  html_document: default
  pdf_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = TRUE)
```

# Introduction

This notebook shows standard quality control steps to test GpC methyltransferase enzyme activity of PEPC enzyme, comparing with the reference (NEB enzyme).

## Data
_Drosophila_ S2 cells were treated with different concentration of the in-house generated GpC methyltransferase enzyme (24 units, 32 units, and 40 units) and NEB enzyme as a reference. All conditions were generated in duplicates. SMF library were sequenced using Illumina NextSeq...P... platform. 

Once obtain the raw sequencing data, Fastq files were aligned to the _Drosophila_ dm6 genome using `QuasR` package. Duplicate reads were removed using `Picard` and `Samtools`. Then, BAM files from replicates of the same treatment condition were merged.

In this example, we subset the BAM files to contain only reads from `chr4` only.


# PEPC vs NEB Enzyme Efficiency

## Load libraries
```{r}
library(QuasR)
library(SingleMoleculeFootprinting)
library(BSgenome.Dmelanogaster.UCSC.dm6)
library(tidyverse)
library(ggplot2)
library(GGally)
library(here)
```

## Create QuasR input file
We will first create QuasR input file, which will be used by QuasR package to locate the BAM files and sample name.
QuasR input file contains two columns: 1) BAM file names, and 2) sample names.
All BAM files for this example are in the folder `/data/dataset_2_S2_PEPCvsNEB`
```{r}
Qinput_df <- data.frame(FileName = c(paste0(here::here(), "/data/dataset_2_S2_PEPCvsNEB/Batch7_24_NO_deDup_merged_chr4.bam"),
                                     paste0(here::here(), "/data/dataset_2_S2_PEPCvsNEB/Batch7_32_NO_deDup_merged_chr4.bam"),
                                     paste0(here::here(), "/data/dataset_2_S2_PEPCvsNEB/Batch7_40_NO_deDup_merged_chr4.bam"),
                                     paste0(here::here(), "/data/dataset_2_S2_PEPCvsNEB/NEB_NO_deDup_merged_chr4.bam")),
                        SampleName = c("Batch7_24_NO_chr4",
                                       "Batch7_32_NO_chr4",
                                       "Batch7_40_NO_chr4",
                                       "NEB_NO_chr4"))

print(Qinput_df)

write.table(Qinput_df, paste0(here::here(), "/analysis/example_2/Qinput_PEPC_batch7_chr4.txt"),
                              col.names = TRUE, row.names = FALSE, sep = "\t", quote = FALSE)

Qinput <- paste0(here::here(), "/analysis/example_2/Qinput_PEPC_batch7_chr4.txt")
```

## 1. Cytosine conversion rate
Here, we will quantify the cytosine conversion rate of each sample, by asking how much of the observed cytosine methylation falls into the expected context (GpC in this case).

Please note that this example only contain data from one chromosome. But in an actual full data set, we can simply write a loop over chromosomes to measure cytosine conversion rates.

```{r}
chr_ConversionRate <- ConversionRate(sampleSheet = Qinput,
                                     genome = BSgenome.Dmelanogaster.UCSC.dm6,
                                     chr = "chr4",
                                     cores = 4)
print(chr_ConversionRate)
```
Now we can plot the conversion rate using bar plot.
```{r}
# plotting
chr_ConversionRate %>% 
  as.data.frame() %>%
  rownames_to_column(var = "sample") %>%
  as.tibble() %>%
  rename(ConversionRate = 2) -> plot_df

plot_df %>% ggplot(aes(x = sample, y = ConversionRate, fill = sample)) +
  geom_bar(stat="identity") +
  xlab("Sample") +
  ylab("Cytosine Conversion Rate") +
  theme(axis.text.x = element_text(size = 14, angle = 45, hjust = 1, vjust = 1),
        axis.text.y = element_text(size = 14),
        axis.title.x = element_text(size=15),
        axis.title.y = element_text(size=15),
        legend.title = element_blank(),
        legend.text = element_text(size = 12),
        legend.position = "right",
        panel.background = element_blank(),
        axis.line = element_line(colour = "black"))

```

What actually happened under the hood of the function `ConversionRate`?

```{r}
# unpack the steps in the ConversionRate function
```


## 2. Correlation of cytosine methylation rate between samples
Now, we can compare methylation rate across samples. We will identify the methylation status for the Cytosines in the genomic context relevant for the experiment (Single enzyme, double enzyme, etc.). The type of the experiment is inferred based on which user-provided substring is found in the SampleName field of the QuasR input file: 

### Call context methylation
```{r}
Qinput <- paste0(here::here(), "/analysis/example_2/Qinput_PEPC_batch7_chr4.txt")
MySample <- suppressMessages(readr::read_delim(Qinput, delim = "\t")[[2]])

genome <- BSgenome.Dmelanogaster.UCSC.dm6

# make cluster
cluObj <- makeForkCluster(nnodes = 4)

# region of interest (in this example we only have chromosome 4)
# chromosome 4 is the fifth on the Dmelanogaster genome (chr2L, chr2R, chr3L, chr3R, chr4)
gr <- GRanges(seqnames = seqnames(Dmelanogaster)[5],
              ranges = IRanges(start = 1, end = seqlengths(Dmelanogaster)[5]),
              strand = "*")

print(gr)

# Call context methylation
Context_Methylation <- CallContextMethylation(sampleSheet = Qinput, 
                                              sample = MySample, 
                                              genome = genome, 
                                              RegionOfInterest = gr, 
                                              returnSM = FALSE,
                                              coverage = 20, 
                                              ConvRate.thr = 0.2,
                                              clObj = cluObj)
```

The Context_Methylation object contain, for each sample, an average methylation level at each cytosine in GC and CG contexts.
```{r}
# GC context
Context_Methylation$DGCHN %>% head() %>% print()

# CG context
Context_Methylation$NWCGW %>% head() %>% print()
```

### Pair plot to see correlation of GC methylation level across samples.

#### helper functions for the plot
```{r}
jet.colors <- grDevices::colorRampPalette(c("#00007F", "blue", "#007FFF", "cyan","#7FFF7F", "yellow", "#FF7F00", "red", "#7F0000"))
 
panel.jet <- function(...) {
  smoothScatter(..., nrpoints=0, add=TRUE, colramp=jet.colors) 
  }

panel.hist <- function(x, ...) {
  usr <- par("usr"); on.exit(par(usr))
  par(usr = c(usr[1:2], 0, 1.5) )
  h <- hist(x, plot = FALSE)
  breaks <- h$breaks; nB <- length(breaks)
  y <- h$counts; y <- y/max(y)
  rect(breaks[-nB], 0, breaks[-1], y, col="grey", ...)
  }

panel.cor <- function(x, y, digits=2, prefix="", cex.cor, ...) {
  usr <- par("usr"); on.exit(par(usr))
  par(usr = c(0, 1, 0, 1))
  r <- abs(cor(x, y, use="pairwise.complete.obs"))
  txt <- format(c(r, 0.123456789), digits=digits)[1]
  txt <- paste(prefix, txt, sep="")
  if(missing(cex.cor)) cex.cor <- 2.5/strwidth(txt)
  text(0.5, 0.5, txt)
  }

```


#### Plot GC methylation
```{r}
Context_Methylation_DGCHN <- Context_Methylation$DGCHN
MethRate_df <- Context_Methylation_DGCHN %>% 
    as.data.frame() %>% 
    select(matches("MethRate"))
  
# remove NAs
MethRate_df <- MethRate_df[complete.cases(MethRate_df),]

# clean column names
colnames(MethRate_df) <- unlist(lapply(colnames(MethRate_df), gsub, pattern = "_NO_chr4_MethRate", replacement = ""))

# plot
p <- pairs(MethRate_df, 
           upper.panel = panel.cor, 
           diag.panel = panel.hist, 
           lower.panel = panel.jet,
           cex.labels = 1)
```

From the histogram, we can see that the average methylation level increases as we increase the PEPC enzyme concentration (24, 32, 40 units). Furthermore, we can see decent correlation levels across samples. The correlation between PEPC enzyme-treated samples and NEB enzyme-treated samples increases following the PEPC enzyme concentration.


#### Plot CG methylation
As a comparison, we will plot the average methylation correlation across samples in the CG context. This we do not expect a correlation, as we did not make a CG enzyme treatment. Indeed, the situation will be different if we perform dual-enzyme treatment.
```{r}
Context_Methylation_NWCGW <- Context_Methylation$NWCGW
MethRate_df <- Context_Methylation_NWCGW %>% 
    as.data.frame() %>% 
    select(matches("MethRate"))
  
# remove NAs
MethRate_df <- MethRate_df[complete.cases(MethRate_df),]

# clean column names
colnames(MethRate_df) <- unlist(lapply(colnames(MethRate_df), gsub, pattern = "_NO_chr4_MethRate", replacement = ""))

# plot
p <- pairs(MethRate_df, 
           upper.panel = panel.cor, 
           diag.panel = panel.hist, 
           lower.panel = panel.jet,
           cex.labels = 1)
```




