---
title: "Transcription factor (TF) binding frequency"
author: "Kasit Chatsirisupachai"
date: "2024-06-26"
output:
  html_document: 
    df_print: paged
  pdf_document: 
    latex_engine: xelatex
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Introduction
As you have learned within the course already, we can use SMF to study protein-DNA bindings, as examplified by RNA polymerase II (Pol II).
In this exercise, we will leverage SMF to study the binding of transcription factors (TF) to DNA. The principle of the analysis is very similar to our Pol II example. Basically, you have to do the context methylation calling. Then, you will perform single-molecule sorting of each read, to assign molecular states (e.g. nucleosome, unbound, TF bound).

Here, we have a dataset of mouse embryonic stem cells (mESCs) "Control" and the mESCs with the transcription factor "REST" knocked out (REST KO). We will study REST bindings at their binding sites comparing between these two condition. These data sets come from one of the papers we sent to you earlier (Sonmezer et al., Mol. Cell, 2021). You can have a look at the Figure 2 in the paper.


## Data
The data sets include the control and the REST K/O samples. The control sample is a targeted amplicon DUAL-enzyme SMF (GpC and CpG) on selected mouse transcription factor binding sites (TFBSs) of mESC TKO cell line. The REST K/O is a targeted amplicon SINGLE-enzyme SMF (only GpC) on TFBSs. SMF library were sequenced in paired-end using Illumina platform. Once obtain the raw sequencing data, Fastq files were aligned to the mouse mm10 genome using `QuasR` package.

## Libraries
First of all, if you haven't done so, don't forget to install the BSgenome package for mouse genome using the below code
```{r}
#BiocManager::install("BSgenome.Mmusculus.UCSC.mm10")
```

Once installed, let's load the necessary packages.
```{r, message = FALSE}
library(QuasR)
library(SingleMoleculeFootprinting)
library(BSgenome.Mmusculus.UCSC.mm10)
library(tidyverse)
library(ggplot2)
```

## Load objects
First, we will load the GRanges object containing the binding sites for REST. 
!!! DON'T FORGET TO CHANGE THE PATH TO YOUR OWN PATH !!!
```{r}
REST_TFBS <- readRDS("/g/krebs/chatsiri/SMF_tutorial/data/dataset_4_TF_cooccupancy/REST_TFBSs.rds")
head(REST_TFBS)
```

Next, we will load also the GRanges object containing the amplicon locations we will be working with.
!!! DON'T FORGET TO CHANGE THE PATH TO YOUR OWN PATH !!!
```{r}
amplicon_REST <- readRDS("/g/krebs/chatsiri/SMF_tutorial/data/dataset_4_TF_cooccupancy/Amplicon_REST_regions.rds")
head(amplicon_REST)
```

Notice that we can locate the binding site of REST within each amplicon region by connecting the `TargetSite` column in the `amplicon_REST` object with the `absolute.idx` column in the `REST_TFBS` object.


#### Question 1: 
How many amplicon regions we are looking at? and What is the average width of these amplicon regions in the `amplicon_REST` object?

Answer:

Hint: use function `length()`, `width()` and `mean()`
```{r}
```


Let's start analysing the data! As always, we need to load Qinput files. In this example, we will have two Qinput files, one for the Control sample and the other for the REST_KO sample. We will analyse each of them separately. The reason for this is that the Control sample was footprinted using both GpC and CpG enzymes, while the REST_KO sample was footprinted only by GpC enzyme.

!!! DON'T FORGET TO CHANGE THE ABSOLUTE PATH IN THE QINPUT FILES TO YOUR OWN PATH. (remember we had this problem during the course) !!!
```{r}
Qinput_Control <- "/g/krebs/chatsiri/SMF_tutorial/analysis/example_4/Qinput_Control.txt"
Qinput_REST_KO <- "/g/krebs/chatsiri/SMF_tutorial/analysis/example_4/Qinput_REST_KO.txt"
```

## One region as an example
We will first take a look at the first amplicon region in a Control sample, as an example.

```{r}
MySample <- suppressMessages(readr::read_delim(Qinput_Control, delim = "\t")[[2]])

genome <- BSgenome.Mmusculus.UCSC.mm10

gr <- amplicon_REST[1]
print(gr)
```

### Call context methylation
```{r}
Methylation <- CallContextMethylation(sampleSheet = Qinput_Control, 
                                      sample = MySample, 
                                      genome = genome, 
                                      RegionOfInterest = gr, 
                                      coverage = 20, 
                                      ConvRate.thr = 0.2, 
                                      returnSM = TRUE, 
                                      clObj = 1)

```

### Plot average methylation
```{r}
# locate REST binding sites
TFBS <- REST_TFBS[REST_TFBS$absolute.idx == gr$TargetSite]

PlotAvgSMF(MethGR = Methylation[[1]],
           RegionOfInterest = gr,
           TFBSs = TFBS)
```


#### Question 2: 
What does each dot in the plot represent? What is the y axis, how does it relate to average methylation level?

Answer: 



### Single-molecule sorting
Similar to the Pol II example previously, we can assign molecular state to each read. However, unlike the 4-bin strategy in the Pol II example, the TF binding is simpler. We will use the 3-bin strategy: the middle bin represents the TFBS and flanking by the other two bins.

From 3 bins, there can be 8 molecular states in total (`000`, `001`, `010`, `011`, `100`, `101`, `110`, and `111`).
We will define 3 TF binding states from these 8 total states. 1) TF-bound: `101`, 2) unbound: `111`, and 3) nucleosome-bound: the other 6 states.

Okay, let's now do the sorting. Remember we need to define the bins. Here, I will use the function `MakeBins` as we previously used, but this time with the `singleTF` BinType. Notice that with TF we don't need to use StrandAware option like in the Pol II example.

```{r}
BinsCoordinates <- MakeBins(RegionsOfInterest = TFBS, BinType = "singleTF", StrandAware = FALSE)
SortedReads <- list(TKO_DE_Control = SortReads(Methylation[[2]]$TKO_DE_Control, BinsCoordinates, coverage = 20))
```

We can define molecular states of TF binding as discussed.
```{r}
TF_states <- list(nucleosome = c("000", "001", "010", "011", "100", "110"),
                  unbound = "111",
                  TF_bound = "101")
```

Finally, we can quantify state frequency of nucleosome, unbound, and TF-bound states.


#### Question 3: 
Have a look at the SortedReads object, how many reads are there in the unbound and TF-bound states? And what are the frequency of unbound and TF-bound states?

Answer:

Hint: use the `StateQuantification` function with the `TF_states` we just defined above as a parameter `states`.
```{r}
```



## REST KO sample
Now, it's time to analyse the REST_KO sample to compare REST binding frequency in this example region.

```{r}
MySample <- suppressMessages(readr::read_delim(Qinput_REST_KO, delim = "\t")[[2]])

genome <- BSgenome.Mmusculus.UCSC.mm10

gr <- amplicon_REST[1]
print(gr)

```

### Call context methylation
Again, we perform call context methylation on the REST_KO sample. As it is a single-enzyme treatment (only GpC), the `CallContextMethylation` function returns two contexts (GpC and CpG) separately.
```{r}
Methylation <- CallContextMethylation(sampleSheet = Qinput_REST_KO, 
                                      sample = MySample, 
                                      genome = genome, 
                                      RegionOfInterest = gr, 
                                      coverage = 20, 
                                      ConvRate.thr = 0.2, 
                                      returnSM = TRUE, 
                                      clObj = 1)
print("GC context")
print(head(Methylation[[1]]$DGCHN))

print("CG context")
print(head(Methylation[[1]]$NWCGW))
```



#### Question 4:
Can you repeat the above average SMF plot that we did for the Control sample on the REST_KO sample? Don't forget to only select for the GC context for plotting.

What do you observe?

Answer:

You can type on the below code chunck for plotting.
```{r}
```


### Single-molecule sorting
Similar to the control sample, we can sort each DNA molecule in different states (nucleosome, unbound, TF-bound).
Again, we only use information from the GC context.
```{r}
BinsCoordinates <- MakeBins(RegionsOfInterest = TFBS, BinType = "singleTF", StrandAware = FALSE)
SortedReads <- list(REST_NO_KO = SortReads(Methylation[[2]]$REST_NO_KO$DGCHN, BinsCoordinates, coverage = 20))
```


#### Question 5: 
Have a look at the SortedReads object, do you observe any REST-bound state?

Answer:

Hint: use the `StateQuantification` function with the `TF_states` as a parameter `states`.
```{r}
```



## Working with multiple REST binding sites
Now, we will simultaneously quantify REST binding states in multiple REST binding sites.

### Control sample
```{r, message = FALSE}
MySample <- suppressMessages(readr::read_delim(Qinput_Control, delim = "\t")[[2]])
genome <- BSgenome.Mmusculus.UCSC.mm10


# this would take some time to run.
SortedReads_Control <- lapply(seq_along(amplicon_REST), function(i) {
  CurrentWindow <- amplicon_REST[i]
  print(paste0("Working on: region ", i))
  
  ExperimentType <- suppressMessages(SingleMoleculeFootprinting::DetectExperimentType(Samples = MySample))
  
  # Call context methylation
  Methylation <- CallContextMethylation(sampleSheet = Qinput_Control,
                                        sample = MySample,
                                        genome = genome,
                                        RegionOfInterest = CurrentWindow,
                                        coverage = 20, 
                                        ConvRate.thr = 0.2,
                                        returnSM = TRUE)
  
  # locate REST binding sites and make bins
  TFBS <- REST_TFBS[REST_TFBS$absolute.idx == CurrentWindow$TargetSite]
  BinsCoordinates <- MakeBins(RegionsOfInterest = TFBS, BinType = "singleTF", StrandAware = FALSE)
  
  # Sort reads
  SortedReads_window <- list(TKO_DE_Control = SortReads(Methylation[[2]]$TKO_DE_Control, BinsCoordinates, coverage = 20))
  
  names(SortedReads_window) <- names(amplicon_REST[i])
  
  return(SortedReads_window)
  })

SortedReads_Control <- unlist(SortedReads_Control, recursive = FALSE)

# Calculate state frequencies
StateFrequency_Control <- Reduce(rbind, lapply(seq_along(SortedReads_Control), 
        function(i) {
            StateQuantification_tbl = StateQuantification(SortedReads = SortedReads_Control[i], states = TF_states)
            StateQuantification_tbl
            }
        ))

StateFrequency_Control <- drop_na(StateFrequency_Control)
print(head(StateFrequency_Control))
```


### REST_KO sample
```{r, message = FALSE}
MySample <- suppressMessages(readr::read_delim(Qinput_REST_KO, delim = "\t")[[2]])
genome <- BSgenome.Mmusculus.UCSC.mm10


# this would take some time to run.
SortedReads_REST_KO <- lapply(seq_along(amplicon_REST), function(i) {
  CurrentWindow <- amplicon_REST[i]
  print(paste0("Working on: region ", i))
  
  ExperimentType <- suppressMessages(SingleMoleculeFootprinting::DetectExperimentType(Samples = MySample))
  
  # Call context methylation
  Methylation <- CallContextMethylation(sampleSheet = Qinput_REST_KO,
                                        sample = MySample,
                                        genome = genome,
                                        RegionOfInterest = CurrentWindow,
                                        coverage = 20, 
                                        ConvRate.thr = 0.2,
                                        returnSM = TRUE)
  
  # locate REST binding sites and make bins
  TFBS <- REST_TFBS[REST_TFBS$absolute.idx == CurrentWindow$TargetSite]
  BinsCoordinates <- MakeBins(RegionsOfInterest = TFBS, BinType = "singleTF", StrandAware = FALSE)
  
  # Sort reads
  SortedReads_window <- list(REST_NO_KO = SortReads(Methylation[[2]]$REST_NO_KO$DGCHN, BinsCoordinates, coverage = 20))
  
  names(SortedReads_window) <- names(amplicon_REST[i])
  
  return(SortedReads_window)
  })

SortedReads_REST_KO <- unlist(SortedReads_REST_KO, recursive = FALSE)

# Calculate state frequencies
StateFrequency_REST_KO <- Reduce(rbind, lapply(seq_along(SortedReads_REST_KO), 
        function(i) {
            StateQuantification_tbl = StateQuantification(SortedReads = SortedReads_REST_KO[i], states = TF_states)
            StateQuantification_tbl
            }
        ))

StateFrequency_REST_KO <- drop_na(StateFrequency_REST_KO)
print(head(StateFrequency_REST_KO))
```


### Plotting
We will now make boxplot to visualise the difference in REST binding between the control and the REST_KO samples.
I preprocessed the plotting dataframe for you in the below chuck of code.
```{r}
StateFrequency_Control %>% filter(State == "TF_bound") %>% select(Freqs) -> Control_binding_freq
StateFrequency_REST_KO %>% filter(State == "TF_bound") %>% select(Freqs) -> REST_KO_binding_freq

plot_df <- data.frame(Control = Control_binding_freq$Freqs, REST_KO = REST_KO_binding_freq$Freqs)
```

#### Question 6: 
Can you finish the plot?

```{r}
```
